Feature 67 forgot password

RignonNoel opened the pull request:
| Q                                                      | R
| ------------------------------------------ | -------------------------------------------
| Type of contribution ?                      | Features
| Tickets (_issues_) concerned               | #67 

---

##### What have you changed ?
 - Implement a generic model to manage action's token
 - Allow user to retrieve his password if forgotten

##### How did you change it ?
 - Refactor `ActivationToken` to take a new attribut `type` and allow multiple action type on the same model
 - Rename `ActivationToken` to `ActionToken`
 - Refactor activation flow to be functionnal after refactor
 - Add an endpoint to create a token to change a password
 - Add an endpoint to change a password if we have a token

Verification :
 -  [x] My branch name respect the standard defined in `CONTRIBUTING.md`
 -  [x] This Pull-Request fully meets the requirements defined in the issue
 -  [x] I added or modified the attached tests

francisjeanneau commented on source/apiVolontaria/apiVolontaria/admin.py:
```diff
@@ -22,6 +22,7 @@ class UserAdmin(admin.ModelAdmin):
 
 
 admin.site.register(models.TemporaryToken)
+admin.site.register(models.ActivationToken)
```
Activation? Isn't this misguiding? I would expect an `activation_token` to be used for account "activation". 

Why not use a subclass of temporary token?

RignonNoel replied:
`ActivationToken` already exist from the feature of user's account activation by email and token.

I wanted to use the same model for ALL token of action by adding a `type` attribute to avoid multiplication of DB table just for one kind of token.

I think it will be great to rename this `ActivationToken` to `ActionToken` but it's a big refactor so i wanted to see with you before.

RignonNoel replied:
Temporary token is not relates to this PR.

francisjeanneau commented on source/apiVolontaria/apiVolontaria/views.py:
```diff
@@ -257,3 +263,129 @@ def post(self, request):
+            user = tokens[0].user
+            try:
+                password_validation.validate_password(password=new_password)
+            except Exception as err:
```
Better to catch the specific exception instead of catching all exceptions.

francisjeanneau commented on source/apiVolontaria/apiVolontaria/views.py:
```diff
@@ -257,3 +263,129 @@ def post(self, request):
+
+        # We have multiple token with the same key (impossible)
+        else:
+            error = _("The system have a problem, please contact us, "
```
The system **has** a problem [...]

francisjeanneau commented on source/apiVolontaria/apiVolontaria/views.py:
```diff
@@ -257,3 +264,130 @@ def post(self, request):
+    serializer_class = serializers.UserBasicSerializer
+
+    def post(self, request):
+        token = request.data.get('token')
```
What happens if I don't provide a token? Should we add a serializer to be sure we get required parameters?

RignonNoel replied:
Actually we have an `HTTP_400_BAD_REQUEST` with this message :

```
{'detail': "\"None\" is not a valid token."}
```
I think it can be good to add a serializer as you propose.

francisjeanneau replied:
Should we implement it now?

RignonNoel commented on source/apiVolontaria/apiVolontaria/managers.py:
```diff
@@ -0,0 +1,22 @@
+
+
+class ActionTokenManager(models.Manager):
+    def filter(self, expired=None, *args, **kwargs):
```
Really good point! I took a moment after i read your review to have a thought about it!

I think it's maybe good to let it as i did because :
 - If you put an other type than a boolean, the filter work and return nothing.
 - If we refactor the system of expiration in the future, this code will be more generic and will not need a refactor too.

We just have (i think) no value to check if the type is a boolean or not in this part of the code.

francisjeanneau replied:
Agreed. Since this is a Model Manager, I think we should put the boolean restriction elsewhere (such as a serializer).

RignonNoel commented on source/apiVolontaria/apiVolontaria/urls.py:
```diff
@@ -16,7 +16,8 @@
 
-from .views import ObtainTemporaryAuthToken, Users, UsersId, UsersActivation
+from .views import ObtainTemporaryAuthToken, Users, UsersId, UsersActivation, \
+    ForgotPassword, ChangePassword
```
But if we need to change the architecture we need to change begin of line with this way.

What about this way ? :

```
from .views import (ObtainTemporaryAuthToken, Users, UsersId, UsersActivation, 
                    ForgotPassword, ChangePassword)
```

I've seen a lot of time this kind of import, i think it's easy to manage and it's the same way we use to check the list of fields in our tests.

francisjeanneau replied:
Yeah I think this is much better! :+1:

RignonNoel commented on source/apiVolontaria/apiVolontaria/urls.py:
```diff
@@ -25,6 +26,17 @@
     ),
+    # Forgot password
+    url(
+        r'^token_password$',
```
Beacause i think this way is a little difficult to know what is the first you need to call :
 1. `reset_password`
 2. `change_password`

But yeah, i'm ok with the fact my urls are really bad, i really don't know what to do to be clear,,,

francisjeanneau replied:
I view it this way:
"I want to reset my password!" -> /reset-password
"So this gives me a token eh... Can I change it now?" -> /change-password
I you do it the other way, it'll be flagrant that it is wrong:
"I forgot my password, but want to change it" -> /change-password
"Oops, it needs a token... How do I get that? Oh, I probably need to reset first!" -> /reset-password

francisjeanneau commented:
Admin panel is broken for ActionTokens:
![image](https://user-images.githubusercontent.com/14813038/36564091-2c33a80c-17ea-11e8-8d9e-58bca334d8e6.png)

francisjeanneau commented:
We need to add a serializer to correctly link the API with the database. These are the issues I found so far.

#### POST /token_password
When User is not found:
```
Internal Server Error: /token_password
Traceback (most recent call last):
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/core/handlers/exception.py", line 41, in inner
    response = get_response(request)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/core/handlers/base.py", line 187, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/core/handlers/base.py", line 185, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
    return view_func(*args, **kwargs)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/views/generic/base.py", line 68, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/rest_framework/views.py", line 489, in dispatch
    response = self.handle_exception(exc)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/rest_framework/views.py", line 449, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/rest_framework/views.py", line 486, in dispatch
    response = handler(request, *args, **kwargs)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/source/apiVolontaria/apiVolontaria/views.py", line 283, in post
    user = User.objects.get(username=request.data["username"])
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/db/models/query.py", line 379, in get
    self.model._meta.object_name
django.contrib.auth.models.DoesNotExist: User matching query does not exist.
[22/Feb/2018 21:22:47] "POST /token_password HTTP/1.1" 500 18483
```
When "username" parameter is missing:
```
Internal Server Error: /token_password
Traceback (most recent call last):
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/core/handlers/exception.py", line 41, in inner
    response = get_response(request)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/core/handlers/base.py", line 187, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/core/handlers/base.py", line 185, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
    return view_func(*args, **kwargs)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/views/generic/base.py", line 68, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/rest_framework/views.py", line 489, in dispatch
    response = self.handle_exception(exc)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/rest_framework/views.py", line 449, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/rest_framework/views.py", line 486, in dispatch
    response = handler(request, *args, **kwargs)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/source/apiVolontaria/apiVolontaria/views.py", line 283, in post
    user = User.objects.get(username=request.data["username"])
KeyError: 'username'
[22/Feb/2018 21:24:06] "POST /token_password HTTP/1.1" 500 18044
```
When I provide an existing user in "username" parameter that already has an ActionToken:
```
Internal Server Error: /token_password
Traceback (most recent call last):
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/db/backends/utils.py", line 65, in execute
    return self.cursor.execute(sql, params)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/db/backends/sqlite3/base.py", line 328, in execute
    return Database.Cursor.execute(self, query, params)
sqlite3.IntegrityError: UNIQUE constraint failed: apiVolontaria_actiontoken.user_id

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/core/handlers/exception.py", line 41, in inner
    response = get_response(request)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/core/handlers/base.py", line 187, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/core/handlers/base.py", line 185, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/views/decorators/csrf.py", line 58, in wrapped_view
    return view_func(*args, **kwargs)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/views/generic/base.py", line 68, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/rest_framework/views.py", line 489, in dispatch
    response = self.handle_exception(exc)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/rest_framework/views.py", line 449, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/rest_framework/views.py", line 486, in dispatch
    response = handler(request, *args, **kwargs)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/source/apiVolontaria/apiVolontaria/views.py", line 297, in post
    user=user,
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/db/models/query.py", line 393, in create
    obj.save(force_insert=True, using=self.db)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/source/apiVolontaria/apiVolontaria/models.py", line 74, in save
    return super(ActionToken, self).save(*args, **kwargs)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/db/models/base.py", line 806, in save
    force_update=force_update, update_fields=update_fields)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/db/models/base.py", line 836, in save_base
    updated = self._save_table(raw, cls, force_insert, force_update, using, update_fields)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/db/models/base.py", line 922, in _save_table
    result = self._do_insert(cls._base_manager, using, fields, update_pk, raw)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/db/models/base.py", line 961, in _do_insert
    using=using, raw=raw)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/db/models/query.py", line 1060, in _insert
    return query.get_compiler(using=using).execute_sql(return_id)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/db/models/sql/compiler.py", line 1099, in execute_sql
    cursor.execute(sql, params)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/db/backends/utils.py", line 80, in execute
    return super(CursorDebugWrapper, self).execute(sql, params)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/db/backends/utils.py", line 65, in execute
    return self.cursor.execute(sql, params)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/db/utils.py", line 94, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/utils/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/db/backends/utils.py", line 65, in execute
    return self.cursor.execute(sql, params)
  File "/home/fjeanneau/dev/nousrire/API-Volontaria/env/lib/python3.6/site-packages/django/db/backends/sqlite3/base.py", line 328, in execute
    return Database.Cursor.execute(self, query, params)
django.db.utils.IntegrityError: UNIQUE constraint failed: apiVolontaria_actiontoken.user_id
[22/Feb/2018 21:25:13] "POST /token_password HTTP/1.1" 500 21754
```

#### POST /change_password
When I omit the required parameter in the request:
```
{
  detail: ""None" is not a valid token."
}
```
So what is the parameter that I need to provide?

When I provide "token" parameter with correct token:
```
{
  detail: ""199f9a68f5b12a0602630b6e12b8089e227edf3f" is not a valid token."
}
```

RignonNoel commented:
I've fix your comments :
 - I created two new serializers to check params of `ChangePassword` and `ResetPassword`
 - I renamed `ForgotPassword` to `ResetPassword`
 - I added some new tests to check empty and forget params

yanicolivier commented on source/apiVolontaria/apiVolontaria/views.py:
```diff
@@ -257,3 +265,144 @@ def post(self, request):
+            return Response(content, status=status.HTTP_201_CREATED)
+
+        else:
+            return Response(status=status.HTTP_201_CREATED)
```
Would it be a good idea to send some content here.  

Could only be something like content = { 'detail': {_('Email succesfully sended')} }

RignonNoel replied:
It can be done but for i wanted to have the same logic as the other call of the API.
Actually we return data only when we have a problem.
